### PR TITLE
Add a prototype and infrastructure for reflecting particle boundaries

### PR DESCRIPTION
**Pull request by sbastrakov**
This is not to be merged directly, but to serve as a basis for @lennertsprenger 's implementation.

### CODE REVIEW

**Comment by sbastrakov** [comment]
Note: there is a problem with `KernelShiftParticles` not working for guards as it always gets neighbors even if not used. I will fix it in `dev` branch

File: include/picongpu/particles/Particles.tpp [hunk 1]
```diff
@@ -359,7 +359,14 @@ namespace picongpu
...
 
-        ParticlesBaseType::template shiftParticles<CORE + BORDER>();
+        shiftBetweenSupercells<CORE + BORDER>();
+    }
+
+    template<typename T_Name, typename T_Flags, typename T_Attributes>
+    template<uint32_t T_area>
+    void Particles<T_Name, T_Flags, T_Attributes>::shiftBetweenSupercells()
+    {
+        ParticlesBaseType::template shiftParticles<T_area>();
```
Comment (psychocoderHPC) [inline include/picongpu/particles/Particles.tpp:369]: Note: offline discussed with @sbastrakov, this code can not be called for an area containing GUARD. The reason is that shift particles is touching neighbor supercells data in all direction (one neighbor is required)

Comment (sbastrakov) [reply include/picongpu/particles/Particles.tpp:369]: Yes, I will fix it in `dev` tomorrow.

**Comment by sbastrakov** [comment]
This is mostly superceded by #3776 , and the parts @lennertsprenger needed for his reflective BC branch he already took. Besides, this PR was always intended for prototyping and not to be merged as is. So i will close it, my branch is still there if needed.

**Pull request closed without merging**